test/test_foreach.py

```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+        if opinfo.supports_alpha_param:
+            alpha = 3 if dtype in torch.testing.get_all_int_dtypes() else 3.14
+            self._regular_binary_test(dtype, op, ref, inputs, is_fastpath, alpha=alpha)
+            self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath, alpha=alpha)
+
+    @skipMeta
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensorlists_fastpath(self, device, dtype, op):
+        for N in N_values:
+            disable_fastpath = op.ref == torch.div and dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+            if op.ref == torch.add and dtype == torch.bool:
+                disable_fastpath = True
+            self._test_binary_op_tensorlists(device, dtype, op, N, True, disable_fastpath)
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensorlists_slowpath(self, device, dtype, op):
+        for N in N_values:
+            self._test_binary_op_tensorlists(device, dtype, op, N, False, False)
+
+    def _test_binary_op_scalar(self, device, dtype, opinfo, N, scalar, is_fastpath, disable_fastpath):
+        n_expected_cudaLaunchKernels = N if disable_fastpath else 1
+        op, ref, inplace_op, inplace_ref = self._get_funcs(opinfo, n_expected_cudaLaunchKernels)
+        inputs = [opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath), scalar]
+        self._regular_binary_test(dtype, op, ref, inputs, is_fastpath)
+        self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath)
+
+    @skipCUDAIfRocm
+    @skipMeta
+    @ops(foreach_binary_op_db)
+    def test_binary_op_scalar_fastpath(self, device, dtype, op):
+        for N in N_values:
+            for i, scalar in enumerate([
+                random.randint(1, 10),
+                1.0 - random.random(),
+                True,
+                complex(1.0 - random.random(), 1.0 - random.random()),
+            ]):
+                disable_fastpath = op.ref == torch.div and dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+                # int scalar
+                if i == 0:
+                    disable_fastpath |= dtype == torch.bool
+                # float scalar
+                if i == 1:
+                    disable_fastpath |= dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+                # bool scalar
+                if i == 2:
+                    disable_fastpath |= dtype == torch.bool
+                    if op.ref in (torch.add, torch.mul):
+                        disable_fastpath = False
+                # complex scalar
+                if i == 3:
+                    disable_fastpath |= dtype not in torch.testing.get_all_complex_dtypes()
+                print(scalar, disable_fastpath)
+                self._test_binary_op_scalar(device, dtype, op, N, scalar, True, disable_fastpath)
+
+    @dtypes(*torch.testing.get_all_dtypes())
```

Comment:
It's probably fine for these ops, but generally ops probably don't support `get_all_dtypes()`, so it's dangerous decorator to use